bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,797 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The entrypoint of the v0 scheme for use by bootstrap agent
+
+use super::peer_networking::{
+    spawn_client, spawn_server, AcceptedConnHandle, ConnToMainMsg,
+    ConnToMainMsgInner, MainToConnMsg, Msg, PeerConnHandle,
+};
+use super::{ApiError, ApiOutput, Config as FsmConfig, Fsm, RackUuid};
+use crate::trust_quorum::RackSecret;
+use derive_more::From;
+use sled_hardware::Baseboard;
+use slog::{debug, error, info, o, warn, Logger};
+use std::collections::{BTreeMap, BTreeSet};
+use std::net::{SocketAddr, SocketAddrV6};
+use std::time::Duration;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::{interval, Instant, MissedTickBehavior};
+use uuid::Uuid;
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    id: Baseboard,
+    addr: SocketAddrV6,
+    time_per_tick: Duration,
+    learn_timeout: Duration,
+    rack_init_timeout: Duration,
+    rack_secret_request_timeout: Duration,
+}
+
+// An error response from a `PeerRequest`
+#[derive(Debug, From)]
+pub enum PeerRequestError {
+    // An `Init_` or `LoadRackSecret` request is already outstanding
+    // We only allow one at a time.
+    RequestAlreadyPending,
+
+    // An error returned by the Fsm API
+    Fsm(ApiError),
+
+    // The peer task shutdown
+    Recv(oneshot::error::RecvError),
+
+    // Failed to send to a connection management task
+    Send(mpsc::error::SendError<PeerApiRequest>),
+}
+
+/// A request sent to the `Peer` task from the `PeerHandle`
+pub enum PeerApiRequest {
+    /// Initialize a rack at the behest of RSS running on the same scrimlet as this Peer
+    InitRack {
+        rack_uuid: RackUuid,
+        initial_membership: BTreeSet<Baseboard>,
+        responder: oneshot::Sender<Result<(), PeerRequestError>>,
+    },
+
+    /// Initialize this peer as a learner.
+    ///
+    /// Return `()` from the responder when the learner has learned its share
+    InitLearner { responder: oneshot::Sender<Result<(), PeerRequestError>> },
+
+    /// Load the rack secret.
+    ///
+    /// This can only be successfully called when a peer has been initialized,
+    /// either as initial member or learner who has learned its share.
+    LoadRackSecret {
+        responder: oneshot::Sender<Result<RackSecret, PeerRequestError>>,
+    },
+
+    /// Inform the peer of currently known IP addresses on the bootstrap network
+    ///
+    /// These are generated from DDM prefixes learned by the bootstrap agent.
+    PeerAddresses(BTreeSet<SocketAddrV6>),
+
+    /// Get the status of this peer
+    GetStatus { responder: oneshot::Sender<Status> },
+
+    /// Shutdown the peer
+    Shutdown,
+}
+
+// A handle for interacting with a `Peer` task
+pub struct PeerHandle {
+    tx: mpsc::Sender<PeerApiRequest>,
```

Comment:
I was going to complain that we don't actually need `mpsc`, but it looks like Tokio doesn't have a `spsc`, so... as you were.